packages/std/src/coins.rs

```diff
@@ -0,0 +1,346 @@
+use std::any::type_name;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::str::FromStr;
+
+use crate::{Coin, StdError, StdResult, Uint128};
+
+/// A collection of coins, similar to Cosmos SDK's `sdk.Coins` struct.
+///
+/// Differently from `sdk.Coins`, which is a vector of `sdk.Coin`, here we
+/// implement Coins as a BTreeMap that maps from coin denoms to amounts.
+/// This has a number of advantages:
+///
+/// - coins are naturally sorted alphabetically by denom
+/// - duplicate denoms are automatically removed
+/// - cheaper for searching/inserting/deleting: O(log(n)) compared to O(n)
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct Coins(BTreeMap<String, Uint128>);
+
+// Casting a Vec<Coin> to Coins.
+// The Vec can be out of order, but must not contain duplicate denoms or zero amounts.
+impl TryFrom<Vec<Coin>> for Coins {
+    type Error = StdError;
+
+    fn try_from(vec: Vec<Coin>) -> StdResult<Self> {
+        let vec_len = vec.len();
+
+        let map = vec
+            .into_iter()
+            .filter(|coin| !coin.amount.is_zero())
+            .map(|coin| (coin.denom, coin.amount))
+            .collect::<BTreeMap<_, _>>();
+
+        // the map having a different length from the vec means the vec must either
+        // 1) contain duplicate denoms, or 2) contain zero amounts
+        if map.len() != vec_len {
+            return Err(StdError::parse_err(
+                type_name::<Self>(),
+                "duplicate denoms or zero amount",
+            ));
+        }
+
+        Ok(Self(map))
+    }
+}
+
+impl TryFrom<&[Coin]> for Coins {
+    type Error = StdError;
+
+    fn try_from(slice: &[Coin]) -> StdResult<Self> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl<const N: usize> TryFrom<[Coin; N]> for Coins {
+    type Error = StdError;
+
+    fn try_from(slice: [Coin; N]) -> StdResult<Self> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl TryFrom<Coin> for Coins {
+    type Error = StdError;
+
+    fn try_from(coin: Coin) -> StdResult<Self> {
+        let mut coins = Coins::default();
+        coins.add(coin)?;
+        Ok(coins)
+    }
+}
+
+impl From<Coins> for Vec<Coin> {
+    fn from(value: Coins) -> Self {
+        value.into_vec()
+    }
+}
+
+impl FromStr for Coins {
+    type Err = StdError;
+
+    fn from_str(s: &str) -> StdResult<Self> {
+        s.split(',')
+            .map(Coin::from_str)
+            .collect::<Result<Vec<_>, _>>()?
+            .try_into()
+    }
+}
+
+impl fmt::Display for Coins {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = self
+            .0
+            .iter()
+            .map(|(denom, amount)| format!("{amount}{denom}"))
+            .collect::<Vec<_>>()
+            .join(",");
+        write!(f, "{s}")
+    }
+}
+
+impl PartialEq<Coin> for Coins {
+    fn eq(&self, other: &Coin) -> bool {
+        self.0.len() == 1 && self.amount_of(&other.denom) == other.amount
+    }
+}
```

Comment:
I'd not implement this. It's basically comparing a collection with an element that is isually not implemented. Is there a use case this is needed for?

Reply:
The use case is that in contracts, you might want to accept only one specific denom and fail if people send more then one (to prevent it from getting stuck), but checking that for a specific amount is probably a bit niche. I also agree it feels a bit weird to compare a container with it's elements, so I will remove it.